### PR TITLE
add mmap callback

### PR DESCRIPTION
**Pull request by gburanov**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 language: go
-go_import_path: github.com/remerge/bolt
+go_import_path: go.etcd.io/bbolt
```
Comment (mbertschler) [inline .travis.yml:2]: why did you change the path back to the original? This means that we need to have a `replace` in go.mod in every project that uses it, for example even in bidder which doesn't directly import bolt. I think that makes it more complicated

File: db.go [hunk 2]
```diff
@@ -326,6 +328,10 @@ func (db *DB) hasSyncedFreelist() bool {
 // mmap opens the underlying memory-mapped file and initializes the meta references.
 // minsz is the minimum size that the new mmap can be.
 func (db *DB) mmap(minsz int) error {
+	if db.MmapCallback != nil {
+		db.MmapCallback()
```
Comment (mbertschler) [inline db.go]: Is just having this once good enough? It seems unlikely but it *could* happen that a new transaction is taken from the pool before it is locked?

Maybe it could still be done with one callback that looks like this:
```go
MmapCallback func(locked bool)
```
That then could be used like
```go
if db.MmapCallback != nil {
    db.MmapCallback(true)
    defer db.MmapCallback(false)
}
```
I see that you start the pool again only after the write, but with big batches the write can take 100ms or even a second. While this write happens we would block bids, but the `mmap` call itself finishes quickly.

**Review by mbertschler — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 1 above)
Comment (gburanov) [reply .travis.yml:2]: So the idea is that the change in our custom fork as small as possible. Otherwise (like we had it in our sarama) - we will spend too much time rebasing from the upstream.

Also this solution is probably temporary, and changing path to github.com/remerge/bolt makes it less temporary.

Yes, ```replace``` is the price you pay for that, but it worth it imho

Comment (gburanov) [reply .travis.yml:2]: we have same solution for aerospike now

File: db.go (hunk 2 above)
Comment (gburanov) [reply db.go]: I do not fully understand your point here. So u want a second callback when the Mmap finish working?

We can do that, but what does it really give us? Yes, you are right, sometimes feed operations (select and select random) will be blocked for the whole write period (100 ms or so).

However this happens quite rarely. Do we really need it? =)

Comment (gburanov) [reply db.go]: on the other size it make code in feeddb easier, so I did it

**Review by maprost — APPROVED** [review]

File: .travis.yml (hunk 1 above)
Comment (mbertschler) [reply .travis.yml:2]: Hmm I don't know, why is it temporary? Do you think this feature will be added in the upstream?
The https://github.com/etcd-io/bbolt#project-status says:
> Bolt is stable, the API is fixed, and the file format is fixed.

I don't think they would add this if just we need it.
Because of this I think it will be pretty much a permanent fork.

Comment (gburanov) [reply .travis.yml:2]: uff... we need to get rid of permanent forks when possible. I will try to suggest something to boltdb maintainers. If it is will the unaccepted in any way - the smaller we have the patch - the easier to maintain it - so still better not to change path

**Comment by gburanov** [comment]
This was approved so I will merge it
